pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,350 @@
+	vif, err := c.vifAllocator.setVifIfNotExist(iif)
+	if err != nil {
+		return err
+	}
+	mc.mfccParent = uint16(vif)
+	return setsockoptMfcctl(c.sockFD, MRT_DEL_MFC, mc)
+}
+
+func (c *MRouteClient) flushMRoute() (err error) {
+	klog.Infof("Clearing multicast routing table entry")
+	return setsockoptVifctl(c.sockFD, MRT_FLUSH, &vifctl{})
+}
+
+func (c *MRouteClient) DeleteMrouteEntryByGroup(group net.IP) (err error) {
+	klog.V(2).Infof("Deleting group %s", group.String())
+
+	mEntries, _ := c.inboundRouteCache.ByIndex(GroupNameIndexName, group.String())
+	for _, route := range mEntries {
+		entry := route.(*inboundMulticastRouteEntry)
+		err := c.delMrouteEntry(net.ParseIP(entry.src), net.ParseIP(entry.group), entry.iif)
+		if err != nil {
+			return err
+		}
+		c.inboundRouteCache.Delete(route)
+	}
+	return nil
+}
+
+func (c *MRouteClient) addOutBoundMrouteEntry(src net.IP, group net.IP, oifs []string) (err error) {
+	klog.InfoS("Adding outbound multicast route", "src", src, "group", group, "oifs", oifs)
+	oifsVif := make([]int, 0)
+	for _, oif := range oifs {
+		oifvif, err := c.vifAllocator.setVifIfNotExist(oif)
+		if err != nil {
+			return err
+		}
+		oifsVif = append(oifsVif, int(oifvif))
+	}
+
+	inboundVif, err := c.vifAllocator.setVifIfNotExist(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+	err = c.addMrouteEntry(src, group, uint16(inboundVif), oifsVif)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MRouteClient) addInboundMrouteEntry(src net.IP, group net.IP, inboundVif uint16) (err error) {
+	klog.InfoS("Adding inbound multicast route", "src", src, "group", group, "inboundVif", inboundVif)
+
+	oif, err := c.vifAllocator.setVifIfNotExist(c.gatewayInterface)
+	if err != nil {
+		return err
+	}
+
+	err = c.addMrouteEntry(src, group, inboundVif, []int{int(oif)})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type inboundMulticastRouteEntry struct {
+	group string
+	src   string
+	iif   string
+}
+
+func getMulticastEntryKey(obj interface{}) (string, error) {
+	entry := obj.(*inboundMulticastRouteEntry)
+	return entry.group + "/" + entry.src + "/" + entry.iif, nil
+}
+
+func groupIndexFunc(obj interface{}) ([]string, error) {
+	entry, ok := obj.(*inboundMulticastRouteEntry)
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{entry.group}, nil
+}
+
+func (c *MRouteClient) getExternalMulticastAddrs() (map[string][]net.IP, error) {
+	interfacesMap := make(map[string]bool)
+
+	for _, iface := range c.ifaceStore.GetAllInterfaces() {
+		for _, ip := range iface.IPs {
+			interfacesMap[ip.String()] = true
+		}
+	}
+
+	ret := make(map[string][]net.IP)
+	l, err := netlink.LinkList()
+	if err != nil {
+		return nil, err
+	} else {
+		for _, link := range l {
+			if link.Attrs().Flags > net.FlagMulticast {
+				addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+				if err != nil {
+					return ret, nil
+				}
+				for _, addr := range addrs {
+					_, ok := interfacesMap[addr.IP.String()]
```

Comment:
Should not we just check if ipv4Addr is a multicast IP? Probably it is that I do not understand what this func is for.